pkg/backingstore/reconciler.go

```diff
@@ -795,3 +862,180 @@ func (r *Reconciler) ReconcilePool() error {
+		newPvc := r.PvcAgentTemplate.DeepCopy()
+		newPvc.Name = pvcName
+		newPvc.Namespace = options.Namespace
+		r.Own(newPvc)
+		util.KubeCreateSkipExisting(newPvc)
+
+		r.PodAgentTemplate.Name = fmt.Sprintf("%s-%s-pod-%s", r.BackingStore.Name, options.SystemName, postfix)
+		r.PodAgentTemplate.Namespace = options.Namespace
+		newPod := r.PodAgentTemplate.DeepCopy()
+		newPod.Spec.Volumes[1].PersistentVolumeClaim.ClaimName = pvcName
+		r.Own(newPod)
+		util.KubeCreateSkipExisting(newPod)
+	}
+	return nil
+}
+
+func (r *Reconciler) isPodinNoobaa(pod *corev1.Pod) bool {
+	for _, host := range *r.HostsInfo {
+		if strings.HasPrefix(host.Name, pod.Name) {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Reconciler) updatePodTemplate() error {
+	if r.Secret.StringData["AGENT_CONFIG"] == "" {
+		return util.NewPersistentError("No AgentConfig - Can't add Missing Agents",
+			fmt.Sprintf("Current Backing store spec is: %q", r.BackingStore.Spec.PVPool))
+	}
+	r.PodAgentTemplate.Spec.Containers[0].Env[1].ValueFrom = &corev1.EnvVarSource{
+		SecretKeyRef: &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: r.Secret.Name,
+			},
+			Key: "AGENT_CONFIG",
+		},
+	}
+	r.PodAgentTemplate.Spec.Containers[0].Image = r.NooBaa.Status.ActualImage
+	if r.NooBaa.Spec.ImagePullSecret == nil {
+		r.PodAgentTemplate.Spec.ImagePullSecrets =
+			[]corev1.LocalObjectReference{}
+	} else {
+		r.PodAgentTemplate.Spec.ImagePullSecrets =
+			[]corev1.LocalObjectReference{*r.NooBaa.Spec.ImagePullSecret}
+	}
+	r.PodAgentTemplate.Labels = map[string]string{"pool": r.BackingStore.Name}
+	return nil
+}
+
+func (r *Reconciler) updatePvcTemplate() error {
+	if r.BackingStore.Spec.PVPool.StorageClass != "" {
+		r.PvcAgentTemplate.Spec.StorageClassName = &r.BackingStore.Spec.PVPool.StorageClass
+	}
+	r.PvcAgentTemplate.Spec.Resources = *r.BackingStore.Spec.PVPool.VolumeResources
+	r.PvcAgentTemplate.Labels = map[string]string{"pool": r.BackingStore.Name}
+	return nil
+}
+
+func (r *Reconciler) deletePvPool() error {
+	podsList := &corev1.PodList{}
+	util.KubeList(podsList, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	for i := range podsList.Items {
+		pod := &podsList.Items[i]
+		util.RemoveFinalizer(pod, nbv1.Finalizer)
+		if !util.KubeUpdate(pod) {
+			r.Logger.Errorf("Pod %q failed to remove finalizer %q",
+				pod.Name, nbv1.Finalizer)
+		}
+	}
+	util.KubeDeleteAllOf(&corev1.Pod{}, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	util.KubeDeleteAllOf(&corev1.PersistentVolumeClaim{}, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	return nil
+}
+
+func (r *Reconciler) upgradeBackingStore(sts *appsv1.StatefulSet) error {
+	r.Logger.Infof("Deleting old statefulset: %s", sts.Name)
+	agentConfig := sts.Spec.Template.Spec.Containers[0].Env[1].Value
```

Comment:
will it always be `Env[1]`? maybe it's better to find by name. 
you can add a util function to find Env by name

Reply:
add function to util